### PR TITLE
Dynamodb backend

### PR DESCRIPTION
**Pull request by seiffert**
This PR adds support for DynamoDB as HA backend to Vault. DynamoDB is a highly scalable, managed service provided as part of the Amazon Web Services. I think it would make a great storage backend for Vault as it requires almost no maintenance.

Due to DynamoDB's conditional write capability, it is possible to implement a distributed locking mechanism on top of it. Therefore, we can use it as a HA backend for Vault.

### Data Model

To map Vault's hierarchical data model to DynamoDB's key/value and document-oriented data model, and still be able to query the backend efficiently, this PR splits the Vault keys into two components: a `path` and a `key`:

* `foo` becomes `Path: "", Key: "foo"`
* `foo/bar` becomes `Path:"foo", Key: "bar"`

This way, the backend's `List` operation can be implemented with a DynamoDB `Query` API call that lists all keys under a given path. In order to list 'sub-folders' (`foo/` in the examples above) as well, a pseudo record `Path: "", Key: "foo/"` is inserted for the 'folder' `foo`.

This data model means that the memory and also computational complexity of `Put` and `Delete` operations is `O(c)` where `c` is the number of path components in the Vault key. To make this more efficient, this PR uses DynamoDB's `BatchWriteItem` API which performs writes in batches of 25.

Locking is done via the same data model with a prefix in the lock entries' keys. This prefix is used to exclude locks from the result of `List` operations. To ensure atomicity, the locking mechanism is implemented via DynamoDB's conditional writes which allow us to only create a lock record if it does not yet exist.

### Configuration

This backend has several configuration options:

* the **DynamoDB table** to use.. Can be configured either via an environment variable `AWS_DYNAMODB_TABLE` or via a `table` option in the server's config file. If this table does not exist yet, it is created during bootstrap.
* the **DynamoDB read capacity** to provision when creating the table. Can be configured wither via an environment varialbe `AWS_DYNAMODB_READ_CAPACITY` or via the `read_capacity` option in the server's config file.
* the **DynamoDB write capacity** to provision when creating the table. Can be configured wither via an environment varialbe `AWS_DYNAMODB_WRITE_CAPACITY` or via the `write_capacity` option in the server's config file.
* the **AWS IAM access key** to authenticate with. Can be configured via an environment variable `AWS_ACCESS_KEY_ID` or via a `access_key` option in the server's config file.
* the **AWS IAM secret key** to authenticate with. Can be configured via an environment variable `AWS_SECRET_ACCESS_KEY` or via a `secret_key` option in the server's config file.
* the **DynamoDB endpoint** to use. Can be configured via an environment variable `AWS_DYNAMODB_ENDPOINT` or via a `endpoint` option in the server's config file.
* the **AWS region** to use. Can be configured via an environment variable `AWS_DEFAULT_REGION` or via a `region` option in the server's config file.
* Whether to **recover** from a system crash. In this scenario, Vault removes a potentially still active lock from DynamoDB. Set this option when all Vault instances were killed / crashed without the chance to clean up the old leadership lock. Enable this mode by setting the environment variable `RECOVERY_MODE` to `1` or via the `recovery_mode` option in the server's config file. Be aware that you should only start *one* server in this mode! After it has been activated, you can start other nodes without the option set.

**Example:**

```
backend "dynamodb" {                            
    table = "my-vault-dynamodb-table"
    region = "eu-west-1"
    advertise_addr = "http://127.0.0.1:8200"
}
```

/cc @mlafeldt

### CODE REVIEW

**Comment by seiffert** [comment]
I'm not sure how the failing tests on Travis can relate to this PR.

**Comment by jefferai** [comment]
@seiffert Thanks for submitting this!

Regarding the failing tests, don't worry about them, they're not at all related to this. Two are in a library that Vault uses, and I need to discuss resolving them with the author, but he's on vacation. The other is in a backend that is only used during unit tests, and while it fails pretty consistently on Travis, it only fails now and again on my own machine which makes it hard to debug and fix. It's nothing to worry about though.

File: physical/dynamodb.go [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+	}
+	writeCapacity, err := strconv.Atoi(writeCapacityString)
+	if err != nil {
+		return nil, fmt.Errorf("invalid write capacity: %s", writeCapacityString)
+	}
+	if writeCapacity == 0 {
+		writeCapacity = DefaultDynamoDBWriteCapacity
+	}
+
+	accessKey, ok := conf["access_key"]
```
Comment (jefferai) [inline physical/dynamodb.go]: Can you add an environment check for `AWS_ACCESS_KEY_ID` here?

File: physical/dynamodb.go [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+	if !ok {
+		accessKey = ""
+	}
+	secretKey, ok := conf["secret_key"]
```
Comment (jefferai) [inline physical/dynamodb.go]: Can you add a check for `AWS_SECRET_ACCESS_KEY` here?

**Comment by jefferai** [comment]
This is really cool, thanks! The code generally looks pretty good to me -- I made two minor requests above.

One other thing: HA relies on being able to write a specific document. If all Vault nodes go down unexpectedly (so they do not clean up properly), will any of them be able to acquire leadership, or will the existence of that record prevent it? If not, I suggest adding an env var/conf flag to be used in case of emergency that basically says "you're the only active node, if the document exists delete it and create a new one". Then you can start one node with that flag one time, then start the other nodes without it.

**Comment by seiffert** [comment]
Thanks for the feedback!
I added explicit checks for the `AWS_*` environment variables. Note that this changes only the precendence of the environment variables vs. the configuration in the config file. The AWS SDK reads those environment variables by default with the `EnvProvider` configured [here](https://github.com/hashicorp/vault/pull/878/files#diff-4e172f54597d9c436085b8be2601c48eR155). By default they are used if no credentials are configured explicitly.

**Comment by jefferai** [comment]
@seiffert Ah, OK. I wasn't aware of that, but even so, we prefer env vars over configuration files so it's better to be explicit about it here, I think.

Any thoughts on the HA scenario I outlined above?

**Comment by seiffert** [comment]
I just tested this scenario. You're right, if Vault is killed without the chance to clean up the lock, new instances can't acquire leadership afterwards. I'm going to add a configuration flag as you described!

**Comment by seiffert** [comment]
@jefferai I've add a flag `recovery_mode`/`RECOVERY_MODE` that causes the DynamoDB backend to delete the lock entry. I tested this with two local vault instances using the same DynamoDB table. After killing (`kill -9) both and re-starting one the normal way, the instance went into `standby` stage. When I started it again with the flag set to `1`, the old lock was removed and the instance acquired leadership.

I'm not sure about the naming of this option. Do you have a better suggestion?

**Comment by jefferai** [comment]
I don't have a better idea about naming, so I think that's fine. Two other asks:

1. Can you update the website documentation (`website/source/docs/config/index.html.md`) with information about the backend?

2. Can you expand the `recovery` option text a bit more to explain the recovery procedure? For instance, it's probably important that only *one* Vault server be started in this mode, rather than setting all of your HA nodes to have it.

**Comment by seiffert** [comment]
I've added some documentation and rebased the branch.

**Comment by jefferai** [comment]
LGTM! Ready to merge it if you think it's complete.

**Comment by seiffert** [comment]
Yes, from my side, we're ready to merge! Thank you!

**Comment by jefferai** [comment]
Done. Thanks!

**Comment by SudBisht** [comment]
Using DynaoDB as HA backend With AWS CLB, but most of the request failing with Request Timeout.
redirect_addr = CLB address on all 3 nodes.
vault version: 0.7.3
@jefferai  is there a way to resolve this issue.

**Comment by SudBisht** [comment]
how we achieve vault HA With DynamoDB: using [vault endpoint](https://www.vaultproject.io/api/system/health.html) for health check in Load Balancer. Only Active node's health will return 200 and health status will be InService. so all traffic will be forwarded to only active node. whenever Active node goes down, as soon as one of the standby node takes the lock, Load Balancer will notice it and mark it as InService and start to redirecting all requests to that instance

**Pull request merged**